### PR TITLE
test(clab): add qemu based clab deployment 

### PR DESCRIPTION
**Pull request by zeeke**
**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> /kind cleanup
> /kind feature
> /kind design
/kind flake
> /kind failing
> /kind documentation
> /kind regression
> /kind example

**What this PR does / why we need it**:

This PR introduce  a containerlab deployment where the cluster is deployed via qEMU and k3s. The main purpose of this work is to enable testing DPDK scenarios with Grout, as qEMU allows emulated IGB network devices.

Current changes still leverages the Grout TAP device mechanism. A subsequent PR will introduce the DPDK workflow a related e2etests.

**Special notes for your reviewer**:

Ref:
- https://github.com/openperouter/openperouter/pull/549
- https://github.com/openperouter/openperouter/pull/753

**Release note**:

```release-note
None
```

**AI Guidelines Acknowledgment**:

- [x] I have reviewed all changes in this PR, including any AI-generated content, and I take full responsibility for its accuracy and correctness.

## Summary by CodeRabbit

- **New Features**
  - Added a QEMU-based Kubernetes and Containerlab environment for end-to-end networking validation.
  - Added automated scenarios covering accelerated underlay connectivity, BGP sessions, EVPN routes, and L3 networking.
  - Added commands to deploy, test, access, clean up, and fully remove the QEMU environment.
  - Added automatic container image loading into the test cluster.
  - Added CI automation to run QEMU end-to-end tests and preserve failure logs.

- **Bug Fixes**
  - Improved failed-test diagnostics with VM, routing, Kubernetes, networking, and hardware details.

- **Documentation**
  - Added architecture, setup, workflow, cleanup, and troubleshooting guidance.

### CODE REVIEW

File: e2etests/pkg/triage/dump.go [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier:Apache-2.0
 
-package tests
+package triage
```
Comment (fedepaol) [inline e2etests/pkg/triage/dump.go:3]: any reason for not having this under pkg?

File: e2etests/tests/common.go [hunk 2]
```diff
@@ -22,3 +25,19 @@ var GroutSupport = ginkgo.Label("grout-support")
...
+
+func dumpIfFails(cs clientset.Interface, additionalNamespaces ...string) {
+	triage.DumpIfFails(cs, triage.Config{
+		ReportPath:           ReportPath,
+		HostMode:             HostMode,
+		GroutMode:            GroutMode,
+		K8sReporter:          K8sReporter,
+		IncludeFRRK8sPods:    true,
+		IncludeFRRContainers: true,
+		IncludePodman:        HostMode,
```
Comment (fedepaol) [inline e2etests/tests/common.go]: please leave a comment here, I find the assignement not straightforward.

File: e2etests/triage/dump.go [hunk 3]
```diff
@@ -20,40 +20,57 @@ import (
...
 )
 
-func dumpIfFails(cs clientset.Interface, additionalNamespaces ...string) {
+// Config controls which environment-specific diagnostics are collected.
+type Config struct {
+	ReportPath           string
+	HostMode             bool
+	GroutMode            bool
+	K8sReporter          *k8sreporter.KubernetesReporter
+	IncludeFRRK8sPods    bool
```
Comment (fedepaol) [inline e2etests/triage/dump.go]: I'd rename these to dump / collectXXX

File: e2etests/tests/validate_session.go [hunk 4]
```diff
@@ -90,16 +64,7 @@ type validationParameters struct {
...
-		evpn, err := frr.EVPNInfo(exec)
-		if err != nil {
-			return err
-		}
-		if !evpn.ContainsType5Prefix(prefix) {
-			return fmt.Errorf("Type-5 route for %s not yet present", prefix)
-		}
-		return nil
-	}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred())
+	validate.Type5RouteExists(exec, prefix)
```
Comment (fedepaol) [inline e2etests/tests/validate_session.go:48]: I'd replace this fully, no need for the wrapper imo

File: e2etests/tests/validate_session.go [hunk 5]
```diff
@@ -46,39 +46,13 @@ func validateFRRK8sSessionForHostSession(name string, hostsession v1alpha1.HostS
...
-			if isRxReceived {
-				continue
-			}
-			return fmt.Errorf("neighbor %s to %s - %s is established but expectedReceivedAF %s not found",
-				parameters.fromName, parameters.toName, parameters.neighborIP, expectedReceivedAF)
-		}
-
-		return nil
-	}, 5*time.Minute, time.Second).ShouldNot(HaveOccurred())
+	validate.SessionWithNeighbor(exec, validate.SessionParameters{
```
Comment (fedepaol) [inline e2etests/tests/validate_session.go]: same here, do we need the wrapper at all?

File: clab/qemu/vm/run_vm.sh
Comment (fedepaol) [inline clab/qemu/vm/run_vm.sh:1]: can we call this file "run.sh" or "run_vm.sh"?

File: clab/qemu/vm/collect-logs.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: Apache-2.0
+#
+# Collects diagnostic logs from the QEMU VM and clab FRR containers.
```
Comment (fedepaol) [inline clab/qemu/vm/collect-logs.sh:4]: what this is for? Ideally, we should be able to collect this info from the test suite when a test fails, keeping an eye on how to run this on multiple environments like a real cluster

File: clab/qemu/vm/setup.sh [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+readonly K3S_VERSION="${K3S_VERSION:-v1.36.4+k3s1}"
+readonly MULTUS_VERSION="${MULTUS_VERSION:-v4.2.1}"
+readonly CNI_PLUGINS_VERSION="${CNI_PLUGINS_VERSION:-v1.9.2-0.20260803142000-012159164d7f}"
+readonly K8S_PORT="${QEMU_K8S_PORT:-6443}"
+
+
+apt-get update
+DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends openssh-client ca-certificates git
+
+curl -Lo /usr/local/bin/kubectl https://dl.k8s.io/release/v1.37.0/bin/linux/amd64/kubectl
```
Comment (fedepaol) [inline clab/qemu/vm/setup.sh]: we should be able to use the local one provided by makefile

File: e2etests/qemu_suite/qemu_scenarios.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+			CIDRs: []string{"100.65.0.0/24"},
+		},
+	},
+}
+
+// --- EVPN accelerated scenarios ---
+
+const testNamespace = "test-clab-l2vni"
+
+var _ = Describe("QEMU scenarios", Ordered, GroutSupport, func() {
```
Comment (fedepaol) [inline e2etests/qemu_suite/qemu_scenarios.go]: can we find a better name for this suite? Like, HWEmulation or something like that?

File: e2etests/qemu_suite/qemu_scenarios.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+		})).To(Succeed())
+
+		By("Verifying host session CIDR in FRR running config")
+		for exec := range routers.GetExecutors() {
+			Eventually(func() error {
+				cfg, err := frr.RunningConfig(exec)
+				if err != nil {
+					return fmt.Errorf("failed to get FRR running config from %s: %w", exec.Name(), err)
+				}
+				if !strings.Contains(cfg, "192.169.10.") {
```
Comment (fedepaol) [inline e2etests/qemu_suite/qemu_scenarios.go]: I don't think this is necessary as we should test high level behavior, corner cases aside

File: e2etests/qemu_suite/suite_test.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+var (
+	nodeLinkConfigPath string
+)
+
+func handleFlags() {
+	flag.StringVar(&executor.Kubectl, "kubectl", "kubectl", "the path for the kubectl binary")
+	flag.StringVar(&ReportPath, "reporterpath", "/tmp", "the path for the reporter")
+	flag.BoolVar(&HostMode, "systemdmode", false, "tells if openperouter is running on the host")
+	flag.BoolVar(&GroutMode, "groutmode", false, "tells if openperouter is running with grout dataplane")
```
Comment (fedepaol) [inline e2etests/qemu_suite/suite_test.go]: do we envision running this also for non grout?

**Review by fedepaol — CHANGES_REQUESTED** [review]

File: e2etests/pkg/triage/dump.go (hunk 1 above)
Comment (zeeke) [reply e2etests/pkg/triage/dump.go:3]: no, moving to e2etests/pkg/triage

File: e2etests/tests/common.go (hunk 2 above)
Comment (zeeke) [reply e2etests/tests/common.go]: I couldn't find any good reason from commit history.
https://github.com/openperouter/openperouter/commit/d1f8b6908e48f5faa2f95b8dd08bdedec19a45d2

Better removing the field and call `dumpPodmanInfo` when in HostMode. the hostmode information is already in the triage.Config struct

File: e2etests/tests/validate_session.go (hunk 5 above)
Comment (zeeke) [reply e2etests/tests/validate_session.go]: it keeps the PR more focused on the subject.
I'll add the wrapper removal (and the changes to the call sites) in a separated commit, to save reviability a little

File: clab/qemu/vm/collect-logs.sh (hunk 6 above)
Comment (zeeke) [reply clab/qemu/vm/collect-logs.sh:4]: most of the information can come from the `e2etests/pkg/triage`. I can drop this script

File: e2etests/qemu_suite/suite_test.go (hunk 10 above)
Comment (zeeke) [reply e2etests/qemu_suite/suite_test.go]: 90% I'd say no, do you prefer to remove the flag? same for HostMode I guess

Comment (fedepaol) [reply e2etests/qemu_suite/suite_test.go]: yeah let's remove them, we can still add them later if needed

File: e2etests/qemu_suite/qemu_scenarios.go (hunk 8 above)
Comment (zeeke) [reply e2etests/qemu_suite/qemu_scenarios.go]: ok for HWEmulation

File: clab/qemu/vm/setup.sh (hunk 7 above)
Comment (zeeke) [reply clab/qemu/vm/setup.sh]: passing the kubectl from the workspace to the VM involved these changes:

https://github.com/openperouter/openperouter/pull/751/changes#diff-c41344c4694855f12d0e9c1014403a476cd319e4299bb24b8efc773a7928ff4aL21-R24

to share the entire repository dir to the clab docker container. before the changes, only the ./clab subfolder was passed

File: clab/qemu/vm/collect-logs.sh (hunk 6 above)
Comment (fedepaol) [reply clab/qemu/vm/collect-logs.sh]: let's drop then

File: clab/qemu/vm/setup.sh [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+echo "Deploying FRR-k8s..."
+"${KUBECTL}" apply -k "/frr-k8s-manifests"
+
+echo "Installing CNI plugins in the VM..."
+ssh_vm sudo bash -s -- "${CNI_PLUGINS_VERSION}" <<'EOF'
+set -euo pipefail
+cni_plugins_version=$1
+
+dnf install -y golang
```
Comment (fedepaol) [inline clab/qemu/vm/setup.sh]: we can defer this (please file an issue) but this logic is the same we have under clab/scripts, it's be good if we can share them

File: e2etests/qemu_suite/qemu_scenarios.go (hunk 8 above)
Comment (fedepaol) [reply e2etests/qemu_suite/qemu_scenarios.go]: can we also rename the file / suite? Sorry if I wasn't clear

File: clab/qemu/vm/cloud-init/user-data [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#cloud-config
```
Comment (fedepaol) [inline clab/qemu/vm/cloud-init/user-data]: General note: I am having second thoughts in putting this under clab.
The clab folder is becoming too bloated with too many things, making it hard to discover where the qemu based env lives (because it's yet another folder together with the nodes configurations).

I'd say we can merge this as it is, but I'll file a followup issue to rearrange the folder.

File: e2etests/tests/validate_session.go (hunk 4 above)
Comment (fedepaol) [reply e2etests/tests/validate_session.go]: still non addressed I think?

File: clab/qemu/vm/setup.sh (hunk 11 above)
Comment (zeeke) [reply clab/qemu/vm/setup.sh]: which part precisely? the CNI binary installation? I didn't find anything similar under clab/scripts

File: clab/qemu/vm/cloud-init/user-data (hunk 12 above)
Comment (zeeke) [reply clab/qemu/vm/cloud-init/user-data]: ok, no hard feelings on the folder position, as long as it contains all the stuff about qemu (vm, scripts, ssh-key, ...)

File: e2etests/tests/validate_session.go (hunk 4 above)
Comment (zeeke) [reply e2etests/tests/validate_session.go]: right. it's a leftover of my work.
I improved the qemu test and it doesn't need `validate.Type5RouteExists` anymore. it can stay as it was before

File: clab/qemu/vm/setup.sh (hunk 11 above)
Comment (fedepaol) [reply clab/qemu/vm/setup.sh]: frrk8s for sure https://github.com/openperouter/openperouter/blob/main/clab/scripts/07-frr-k8s-setup.sh
The same script deploys cni binaries too.

File: clab/qemu/vm/cloud-init/user-data (hunk 12 above)
Comment (fedepaol) [reply clab/qemu/vm/cloud-init/user-data]: let's keep it as it is for now

File: clab/qemu/deploy-clab.sh [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+set -euo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+CLAB_DIR="${SCRIPT_DIR}/.."
+CLAB_NAME="${CLAB_NAME:-kind}"
+
+# shellcheck disable=SC1091
+source "${CLAB_DIR}/common.sh"
+
+if sudo containerlab inspect --name "${CLAB_NAME}" &>/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (zeeke) [reply clab/qemu/deploy-clab.sh]: not addressing. the fix becomes way too complicated to read to justify the problem:

```

if topology=$(sudo containerlab inspect --name "${CLAB_NAME}" --format json 2>/dev/null); then
    if ! jq -e --arg node "clab-${CLAB_NAME}-pe-kind-control-plane" \
        'any(.[][]; .name == $node)' <<<"${topology}" >/dev/null; then
        echo "Topology '${CLAB_NAME}' conflicts with QEMU: missing 'pe-kind-control-plane' node." >&2
        exit 1
    fi
    echo "Containerlab topology '${CLAB_NAME}' is already deployed, skipping deploy."
else
 ...
```

File: clab/qemu/Makefile [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+qemu-load-image: ## Load the container image into the QEMU VM's k3s
+	$(QEMU_DIR)/vm/load-image.sh $(IMG)
+
+.PHONY: qemu-clab
+qemu-clab: qemu-image kubectl ## Deploy containerlab topology and bootstrap k3s in the QEMU guest
+	KUBECONFIG_PATH=$(KUBECONFIG_PATH) \
+	  $(QEMU_DIR)/deploy-clab.sh
+
+.PHONY: qemu-deploy
+qemu-deploy: qemu-clab qemu-load-image deploy-controller ## Full QEMU environment deploy
```
[the remark this answers is not in the document]
Comment (zeeke) [reply clab/qemu/Makefile]: no need to support parallel makefile execution for this kind of rules

File: e2etests/pkg/triage/dump.go [hunk 15]
```diff
@@ -230,6 +251,48 @@ func logFileFor(base string, kind string) (*os.File, error) {
 	return f, nil
 }
 
+// dumpNodePCIInfo collects PCI device details and their bound kernel drivers
+// from every Kubernetes node. The node executor enters the node's namespaces,
+// so this captures the node rather than the node-exec helper container.
+func dumpNodePCIInfo(cs clientset.Interface, basePath, testName string) {
```
Comment (fedepaol) [inline e2etests/pkg/triage/dump.go]: just realized, should this go to the inspect tool https://github.com/openperouter/openperouter/tree/main/tools/inspect ?
Can you file an issue to extend it?

File: clab/qemu/vm/setup.sh (hunk 11 above)
Comment (fedepaol) [reply clab/qemu/vm/setup.sh]: please file an issue for this too

**Comment by fedepaol** [comment]
I'd move this forward with the following leftovers:

https://github.com/openperouter/openperouter/issues/778

plus please file two issues for sharing the frrk8s deployment / cni logic and another one for enhancing the inspect tool with the grout related informations.

Long term (or even not so long) we'll be using the inspect tool to collect all the information related to failed tests, so we can keep it reliable and dogfood ourselves.

**Comment by zeeke** [comment]
> I'd move this forward with the following leftovers:
> 
> #778
> 
> plus please file two issues for sharing the frrk8s deployment / cni logic and another one for enhancing the inspect tool with the grout related informations.
> 
> Long term (or even not so long) we'll be using the inspect tool to collect all the information related to failed tests, so we can keep it reliable and dogfood ourselves.

opened
- https://github.com/openperouter/openperouter/issues/779
- https://github.com/openperouter/openperouter/issues/780
